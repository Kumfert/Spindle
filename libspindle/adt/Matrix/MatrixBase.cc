//
// MatrixBase.h -- a base class for all matrix/graph classes
//
//  $Id: MatrixBase.cc,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//


#include "spindle/MatrixBase.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

bool 
MatrixBase::resize( const int newNCols, const int newNRows, const int newNNonZeros ) {
  
  colPtr.resize( newNCols + 1 );
  rowIdx.resize( newNNonZeros );
  if ( diagPtr.notNull() ) {
    if ( diagPtr.size() == nCols ) {
      diagPtr.resize( newNCols );
    } else { 
      diagPtr.resize( 0 );
    }
  }

  if ( realValues.notNull() ) {
    if ( realValues.size() == nNonZeros ) {
      realValues.resize( newNNonZeros );
    } else {
      realValues.resize( 0 );
    } 
  }
  
  if ( imgValues.notNull() ) {
    if ( imgValues.size() == nNonZeros ) {
      realValues.resize( newNNonZeros );
    } else {
      realValues.resize( 0 );
    }
  }

  nRows = newNRows;
  nCols = newNCols; 
  nNonZeros = newNNonZeros;
  return true;
}

bool
MatrixBase::setDiagPtr() {
  bool recomputeDiagPtr = false; 
  const int n = nCols;
  if ( diagPtr.size() != n ) {
    // there is no diagPtr, so create it
    diagPtr.resize( n );
    recomputeDiagPtr = true;
  }
  // if we aren't already sure that we must recompute
  if (!recomputeDiagPtr) { // test all values
    for(int i=0; i<n; ++i ) {
      if (rowIdx[diagPtr[i]] != i) {
        recomputeDiagPtr = true;
        break;
      }
    }
  }
  // Now recompute if neccessary
  if (recomputeDiagPtr) {
    const int* colptr = colPtr.lend();
    const int* rowidx = rowIdx.lend();
    for(int i=0; i<n ;i++) {
      for(int j=colptr[i];j<colptr[i+1];j++){
        diagPtr[i]=j;
        if(rowidx[j] == i) {
          break;
        } else if (rowidx[j] > i) {
          // error, should have explicit self edges
          return false;
        }
      }
    }
  }
  return true;
}
