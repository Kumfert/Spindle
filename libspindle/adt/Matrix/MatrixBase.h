//
// MatrixBase.h -- a base class for all matrix/graph classes
//
//  $Id: MatrixBase.h,v 1.2 2000/02/18 01:31:50 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_MATRIX_BASE_H_
#define SPINDLE_MATRIX_BASE_H_

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

SPINDLE_BEGIN_NAMESPACE 

class MatrixUtils;

class MatrixBase { 
friend class MatrixUtils;

protected:
  int nRows;
  int nCols;
  int nNonZeros;

  SharedArray<int> colPtr;
  SharedArray<int> rowIdx;
  SharedArray<int> diagPtr;
  SharedArray<double> realValues;
  SharedArray<double> imgValues;

public:

  bool resize( const int newNCols, const int newNRows, const int newNNonZeros );
  int queryNCols() const { return nCols; }
  int queryNRows() const { return nRows; }
  int queryNNonZeros() const { return nNonZeros; }

  const SharedArray<int>& getColPtr() const { return colPtr;  }
  const SharedArray<int>& getRowIdx() const { return rowIdx;  }
  const SharedArray<int>& getDiagPtr() const { return diagPtr;  }
  const SharedArray<double>& getRealValues() const { return realValues; }
  const SharedArray<double>& getImgValues() const { return imgValues;  }
  
  SharedArray<int>& getColPtr() { return colPtr; }
  SharedArray<int>& getRowIdx() { return rowIdx; }
  SharedArray<int>& getDiagPtr() { return diagPtr; }
  SharedArray<double>& getRealValues() { return realValues; }
  SharedArray<double>& getImgValues() { return imgValues;  }

  bool setDiagPtr();
};

SPINDLE_END_NAMESPACE

#endif
