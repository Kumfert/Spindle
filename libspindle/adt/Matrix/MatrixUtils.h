//
// MatrixUtils.h -- a collection of static methods for manipulating matrices
//
//  $Id: MatrixUtils.h,v 1.2 2000/02/18 01:31:50 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_MATRIX_UTILS_H_
#define SPINDLE_MATRIX_UTILS_H_

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class MatrixBase;
class GraphBase;

class MatrixUtils { 
public:
  enum { STRICT_LOWER = -2, LOWER = -1, 
	 DIAGONAL = 0, UPPER = 1, 
	 STRICT_UPPER = 2 } triangularOption;

  static bool isSymmetric( const MatrixBase* matrix, double tol = -1.0 );
  static bool isHermitian( const MatrixBase* matrix, double tol = -1.0 )
  { return isSymmetric( matrix, tol ); }
  static bool isTriangular( const MatrixBase* matrix, int triangularOption );

  static bool makeDiagonalsExplicit( MatrixBase* matrix );
  static bool makeSymmetric( MatrixBase* matrix );
  static bool makeHermitian( MatrixBase* matrix ) 
  { return makeSymmetric( matrix ); }
  static bool makeTriangular( MatrixBase* matrix, int triangularOption );
  static bool makeTranspose( MatrixBase* matrix );
  static bool makeTranspose( MatrixBase* matrix, SharedArray<int>& Tmap ); // save transpose map
  static bool deepCopy( const MatrixBase* src, MatrixBase* dest );

  static bool metrics( MatrixBase* matrix, double* answers, int nAnswers );

};

SPINDLE_END_NAMESPACE

#endif
