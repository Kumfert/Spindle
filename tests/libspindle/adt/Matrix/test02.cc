//
// testMatrixUtils_1.cc --tests basic MatrixBase operations
//
//
//
//

#include <iostream.h>
#include <iomanip.h>
#include "spindle/MatrixBase.h"
#include "spindle/MatrixUtils.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

//
// This is a simple derived class to set some of the information
// conveniently.  "MatrixBase" wasn't designed to be used directly.
//
class MatrixTest : public MatrixBase { 
public: 
  void setNRows( int nrows ) { nRows = nrows; }
  void setNCols( int ncols ) { nCols = ncols; }
  void setNNZ( int nnz ) { nNonZeros = nnz; }
};


// this is a standard 3x3 grid using a 5pt. stencil
struct MatrixType { 
  int nRows;
  int nCols;
  int nnz;
  int colPtr[10];
  int rowIdx[33];
};

static struct MatrixType Matrix1 = { 9, 9, 33, { 0, 3, 7, 10, 14, 19, 23, 26, 30, 33 }, 
				  { 0, 1, 3,
				    0, 1, 2, 4, 
				    1, 2, 5,
				    0, 3, 4, 6,
				    1, 3, 4, 5, 7, 
				    2, 4, 5, 8, 
				    3, 6, 7, 
				    4, 6, 7, 8,
				    5, 7, 8 }};


// this is the same grid without diagonal elements
static struct MatrixType Matrix2 = { 9, 9, 24, { 0, 2, 5, 7, 10, 14, 17, 19, 22, 24 },
				     { /*0,*/ 1, 3,
					      0, /*1,*/ 2, 4, 
					      1, /*2,*/ 5,
					      0, /*3,*/ 4, 6,
					      1, 3, /*4,*/ 5, 7, 
					      2, 4, /*5,*/ 8, 
					      3, /*6,*/ 7, 
					      4, 6, /*7,*/ 8,
					      5, 7, /*8,*/
					      0, 1, 2, 3, 4, 5, 6, 7, 8 }}; // junk to fill up the array

int main() {
  MatrixTest mat1, mat2;

  // now set up mat1
  mat1.setNRows( Matrix1.nRows );
  mat1.setNCols( Matrix1.nCols );
  mat1.setNNZ( Matrix1.nnz );
  mat1.getColPtr().borrow( Matrix1.colPtr, Matrix1.nCols + 1 );
  mat1.getRowIdx().borrow( Matrix1.rowIdx, Matrix1.nnz );

  if ( ! MatrixUtils::isSymmetric( &mat1 ) ) {
    cout << "Warning, MatrixUtils incorrectly labeled mat1 as unsymmetric " << endl;
  }

  // now set up mat2
  mat2.setNRows( Matrix2.nRows );
  mat2.setNCols( Matrix2.nCols );
  mat2.setNNZ( Matrix2.nnz );
  mat2.getColPtr().borrow( Matrix2.colPtr, Matrix2.nCols + 1 );
  mat2.getRowIdx().borrow( Matrix2.rowIdx, Matrix2.nnz );

  // finally get mat2 to call setAdjSelf
  if ( ! MatrixUtils::makeDiagonalsExplicit( &mat2 ) ) {
    cout << "MatrixUtils::makeDiagonalsExplicit( MatrixBase* )" << endl;
  }
  
  // since mat2 borrowed from Matrix2.  The arrays should be different
  if ( mat2.getColPtr().lend() == Matrix2.colPtr ) {
    cout << "Warning, mat2 internal data (which was modified) matches unshared external data." << endl;
  }

  // verify that mat1 and mat2 now have the same number of non-zeros
  if ( mat1.queryNNonZeros() != mat2.queryNNonZeros() ) {
    cout << "Warning mat1 and mat2 do not have same number of non-zeros." << endl;
  }

  // finally print out both mat1 and mat2 and compare
  { 
    const int* colPtr1 = mat1.getColPtr().lend();
    const int* colPtr2 = mat2.getColPtr().lend();
    const int nCols = mat1.queryNCols();
    cout << " ColPtr1   ColPtr2 " << endl;
    for( int i=0 ; i<=nCols; ++i ) {
      cout << setw(4) << i << setw(4) << colPtr1[i] << setw(4) << colPtr2[i] << endl;
    }
  }

  {
    const int* rowIdx1 = mat1.getRowIdx().lend();
    const int* rowIdx2 = mat2.getRowIdx().lend();
    const int nnz = mat2.queryNNonZeros();
    cout << " RowIdx1   RowIdx2 " << endl;
    for( int i=0 ; i<nnz; ++i ) {
      cout << setw(4) << i << setw(4) << rowIdx1[i] << setw(4) << rowIdx2[i] << endl;
    }
  }

  // now make mat2 strictly upper triangular
  if ( ! MatrixUtils::makeTriangular( &mat2, MatrixUtils::STRICT_UPPER ) ) {
    cout << " Warning: trouble making triangular " << endl;
  }

  { 
    const int* colPtr = mat2.getColPtr().lend();
    const int* rowIdx = mat2.getRowIdx().lend();
    const int nCols = mat2.queryNCols();
    // create temporary vector 
    SharedArray<int> temphead;
    temphead.import( colPtr, nCols );

    cout << "Spy of the matrix" << endl;
    for( int i=0; i<nCols; ++i ) { 
      cout << setw(4) << i << ": ";
      for( int j=0; j<nCols; ++j ) {
	if ( (temphead[j] < colPtr[j+1]) && (rowIdx[ temphead[j] ] == i )) {
	  cout << setw(4) << j;
	  temphead[j]++;
	} else { 
	  cout << setw(4) << "";
	}
      }
      cout << endl;
    }
  }



  // now make mat2 symmetric 
  if ( ! MatrixUtils::makeSymmetric( &mat2 ) ) {
    cout << " Warning: trouble making mat2 symmetric again" << endl;
  }

  { 
    const int* colPtr = mat2.getColPtr().lend();
    const int* rowIdx = mat2.getRowIdx().lend();
    const int nCols = mat2.queryNCols();
    // create temporary vector 
    SharedArray<int> temphead;
    temphead.import( colPtr, nCols );

    cout << "Spy of the matrix" << endl;
    for( int i=0; i<nCols; ++i ) { 
      cout << setw(4) << i << ": ";
      for( int j=0; j<nCols; ++j ) {
	if ( (temphead[j] < colPtr[j+1]) && (rowIdx[ temphead[j] ] == i )) {
	  cout << setw(4) << j;
	  temphead[j]++;
	} else { 
	  cout << setw(4) << "";
	}
      }
      cout << endl;
    }
  }
}
    


