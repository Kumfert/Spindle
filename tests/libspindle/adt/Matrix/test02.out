 ColPtr1   ColPtr2 
   0   0   0
   1   3   3
   2   7   7
   3  10  10
   4  14  14
   5  19  19
   6  23  23
   7  26  26
   8  30  30
   9  33  33
 RowIdx1   RowIdx2 
   0   0   0
   1   1   1
   2   3   3
   3   0   0
   4   1   1
   5   2   2
   6   4   4
   7   1   1
   8   2   2
   9   5   5
  10   0   0
  11   3   3
  12   4   4
  13   6   6
  14   1   1
  15   3   3
  16   4   4
  17   5   5
  18   7   7
  19   2   2
  20   4   4
  21   5   5
  22   8   8
  23   3   3
  24   6   6
  25   7   7
  26   4   4
  27   6   6
  28   7   7
  29   8   8
  30   5   5
  31   7   7
  32   8   8
Spy of the matrix
   0:        1       3                    
   1:            2       4                
   2:                        5            
   3:                    4       6        
   4:                        5       7    
   5:                                    8
   6:                                7    
   7:                                    8
   8:                                     
Spy of the matrix
   0:        1       3                    
   1:    0       2       4                
   2:        1               5            
   3:    0               4       6        
   4:        1       3       5       7    
   5:            2       4               8
   6:                3               7    
   7:                    4       6       8
   8:                        5       7    
